QUARKUS-3230 Cooperative write batching

franz1981 opened the pull request:
This is proposing a pattern which reactive folks are pretty used to, see https://akarnokd.blogspot.com/2015/05/operator-concurrency-primitives_11.html

What it solves?
- concurrent writers won't compete awaiting each others to acquire the `lock`, but elect a "combiner" thread which drain their request
- a single thread (light contention or fast disks) won't use the in-flight concurrent queue, but would pay the cost of 2 more atomic ops (increment/decrement) of the shared write counter

What it doesn't solve (and can make things worse, actually!):
- a slow disk means that combiner thread can keep on accumulating writes from other threads, going OOM (a reentrant lock won't do it, unless the caller is a virtual thread, than can still happen on the wait list of the lock)
- it doesn't batch writes/flushes (which can be very beneficial in case flush implies fsync - which can be deferred at the end of a batch)

The second thing could be achieved using a slightly different pattern which doesn't optimize for a single-writer thread/light concurrency as this (see https://github.com/eclipse-vertx/vert.x/blob/04d4d65fb47a3db4cd8a320795a397f1ae678173/src/main/java/io/vertx/core/net/impl/pool/CombinerExecutor.java for the alternative pattern).

What's interesting of this PR is that the shared counters can be used to track in-flight still not completed requests and can be used to perform some form of adapative back-pressure, eg

- if the number of in-flight requests (in bytes or number of items) exceed a threshold, the write request can bring the current thread as a parker, and park awaiting its write request to be completed
- if it happens, the combiner will execute it and unpark it

@geoand  @dmlloyd

geoand commented on src/main/java/org/jboss/logmanager/handlers/WriterHandler.java:
```diff
@@ -19,10 +19,10 @@
-import java.io.Closeable;
-import java.io.Flushable;
-import java.io.Writer;
+import java.io.*;
```
No start imports please

geoand replied:
Oops, I realized this is not the Quarkus repo, so forget about it :)

jamezp replied:
It's okay, I don't want star imports here either :)

jamezp commented on src/main/java/org/jboss/logmanager/handlers/WriterHandler.java:
```diff
@@ -59,25 +73,66 @@ protected void doPublish(final ExtLogRecord record) {
+        }
+    }
+
+    private boolean doWrite(ExtLogRecord record, String formatted) throws IOException {
```
The return value is never used, is that intended?

franz1981 replied:
It was like that in the original version do the code but I can change it

jamezp replied:
No need. It won't hurt anything, just noticed it is all :)

jamezp commented on src/main/java/org/jboss/logmanager/handlers/WriterHandler.java:
```diff
@@ -59,25 +73,66 @@ protected void doPublish(final ExtLogRecord record) {
+            return true;
+        }
+        preWrite(record);
+        final Writer writer = this.writer;
+        if (writer == null) {
+            return true;
+        }
```
Is there a reason for this? It's only used in the lock.

franz1981 replied:
Same here, the code was refactored from ye original code

jamezp reviewed (changes requested):
I need to think about this. The OOME is a concern for me when a lot of logging happens with slow I/O. This handler is used for writing to various streams.

franz1981 commented:
@jamezp Re the OOM I can implement different backpressure strategies. The simpler one is the more punitive, where an incoming producer is blocked if it's write made te backlog to become too big, and wait till the while backlog is cleared up.

Others could employ other mechanisms (using semaphore or other lazy counters), more fair or unfair, to block or slow down producers. 
Let me know If you want me to commit a draft of one of these, but I think we need to make user the original issue is fixed with this already. The backpressure mechanism is to make this prod ready, I think

franz1981 commented:
@jamezp I've implemented the simplest backpressure strategy, somehow similar to what it happen today (let's say)

eg right now everyone battle for the same lock, get enqueued as a waiter and awaken again when got its chance to enter the lock. It means that if a Thread is the very last to enter (assuming a fair enqueue, for simplicity), it will take all the previous and the current requests before being able to make it progress.

What I've implemented is exactly the same behaviour: if the number of waiters become "too much" (now hardcoded), a new incoming one would block till its request is completed.

dmlloyd commented:
How does this compare to using the async handler?

franz1981 commented:
@dmlloyd this one doesn't need any background thread, it perform everything in foreground, electing s different "combiner" everytime ie who execute others concurrent actions

jamezp commented:
TBH this makes me a little nervous. We introduce a queue which in theory could get overloaded and cause issues. If we feel performance is an issue, the `AsyncHandler` could easily be used. If we want to use something like this, we should have a way to either block or discard messages like the `AsyncHandler` already does. 

If the queue becomes overloaded we have potential OOME issues and we can't log it, because it happens in logging.

franz1981 commented:
I have already implemented a backpressured strategy @jamezp which allow a default max level of concurrency which is exactly the same of a reentrant lock.

jamezp commented:
Sorry about that @franz1981. I see the blocking/back-pressure now. Not sure how I missed that.

I'm not necessarily opposed to this. It does seem complicated for what appears to be a simple I/O write, but I understand why. I'd like to get @dmlloyd's opinion here too as he understands the Quarkus side much better than I do.

jamezp commented:
Just saw a question asking for updates on this in the Quarkus JIRA :)

My thought is this seems like a lot of added complication and I'm not sure what we gain. Do we have a measurement of how much this improves performance? While much more naive, the `AyncHandler` would likely have similar performance I'd think. In either case we don't get away from a synchronous write.

That said, I'd like @dmlloyd's opinion too.

dmlloyd commented:
What about only batching flushes? Something like this in the write path:

```java
lock.lock();
try {
   write();
   if (autoFlush && ! lock.hasQueuedThreads()) {
       // nobody else is waiting to write, so flush
       flush();
   }
} finally {
   lock.unlock();
}
```
Then only the last waiter will flush, effectively batching them. Relying on `hasQueuedThreads` might be fragile though for various reasons; we might need some kind of "reliable" version of that (for example keeping a separate wait counter or something).

This avoids additional data structures and avoids the disadvantage that the async handler has that caller info may have to be materialized. It would work as a general performance improvement I think, just reducing the total number of flushes.
